librad/src/git/p2p/server.rs

```diff
@@ -213,17 +213,32 @@ impl UploadPack {
                 let mut stdout = child.stdout.take().unwrap().compat();
 
                 send.write_all(UPLOAD_PACK_HEADER).await?;
-                futures::try_join!(futures::io::copy(&mut stdout, &mut send), child.wait())
-                    .and_then(|(_, status)| {
-                        if !status.success() {
-                            Err(io::Error::new(
-                                io::ErrorKind::Other,
-                                format!("upload-pack ls exited non-zero: {:?}", status),
-                            ))
-                        } else {
-                            Ok(())
-                        }
-                    })
+                let res =
```

Comment:
Alright, trying to spell out what we want here:

* Drive `child` to completion
    
   There is `kill_on_drop`, but that's not a guarantee. That is, if `child` never terminates, the `server` never terminates. Applying some kind of backpressure on the creation of new git streams would not be a bad idea, but that's beyond the scope of this patch.

* Drive `copy(&mut stdout, &mut send)` to completion.

   Same thing: we need to drain `stdout` of the process.

* Drive `recv`, such that it receives either EOF or one byte of data (which, technically, would be a protocol violation).

* Drop all futures, or none at all


So, I think the `try_join` is good. The `copy(&mut recv, &mut devnull)` would continue to receive if the other end is a tarpit. If the other end is a tarpit, however, `receiving_unknown_size()` will be false, so dropping `recv` SHOULD be fine (in that, the stream gets deregistered).

This means that the previous attempt was better, except that the buffer should be 1 byte in length. If that 1 byte is not zero after `AsyncRead::poll_read`-ing once, this function should call the bad manners police immediately with the remote peer's id card.

I'm ok to merge this as-is for now, and follow up with a more elaborate thing.